ES|QL dense vector field type support

carlosdelest opened the pull request:
Support `dense_vector` field type. This is the first step to allow `kNN` queries and having `dense_vector` as a first class citizen in ES|QL

This allows a mapping that has `dense_vector` like the following:

```
{
  "properties": {
    "id": {
      "type": "long"
    },
    "vector": {
      "type": "dense_vector",
      "similarity": "l2_norm"
    }
  }
}
```

To be retrieved via ES|QL:

```esql
FROM dense_vector
| KEEP id, vector
| SORT id
```

```
id   | vector
0    | [1.0, 2.0, 3.0]
1    | [4.0, 5.0, 6.0]
```

For now, just `float` element types are allowed. There will be a similar work in order to allow for `byte` and `bit` element types, but I wanted to review this implementation first to ensure it's in line with what we need.

Both indexed / not indexed types and `synthetic` source is supported.

Support for CSV tests has been added. For now CSV tests are simple, we can expand on these and also support additional operations on `dense_vector` field types in subsequent PRs. An integration test has been added to test extensively on different index options and doc storage structure.

`dense_vector` field type is under a feature flag, as this will require follow up work.

Commit:
Register dense vector field type

Commit:
Add first version of BlockDocValuesReader for dense_vector

Commit:
Fixed BlockDocValuesReader to work with all dense_vector types

Commit:
Merge remote-tracking branch 'origin/main' into feature/esql_dense_vector_support

# Conflicts:
#	x-pack/plugin/esql-core/src/main/java/org/elasticsearch/xpack/esql/core/type/DataType.java
#	x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/action/EsqlCapabilities.java
#	x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/action/ResponseValueUtils.java
#	x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/planner/LocalExecutionPlanner.java

Commit:
Small test fixes

Commit:
Fix casting

Commit:
Improve testing to add random indexing, and fix similarity

Commit:
Add index = false support

Commit:
Spotless

Commit:
Synthetic source testing

Commit:
Add CSV tests and necessary infra for dense_vector field type

Commit:
Make CSV test loader to use numbers when there are multivalued numeric fields

Commit:
Fixed test mapping to avoid parsing errors when numbers are used for multivalued fields

Commit:
Avoid non-float dense vector field types for now

carlosdelest commented on server/src/main/java/org/elasticsearch/index/mapper/BlockDocValuesReader.java:
```diff
@@ -504,6 +506,80 @@ public String toString() {
         }
     }
 
+    public static class DenseVectorBlockLoader extends DocValuesBlockLoader {
```
Added a BlockLoader for dense vectors, that uses `FloatVectorValues` to retrieve indexed vector data.

Commit:
[CI] Auto commit changes from spotless

carlosdelest commented on server/src/main/java/org/elasticsearch/index/mapper/vectors/DenseVectorFieldMapper.java:
```diff
@@ -2311,6 +2329,20 @@ int getVectorDimensions() {
+        @Override
+        public BlockLoader blockLoader(MappedFieldType.BlockLoaderContext blContext) {
+            if (elementType != ElementType.FLOAT) {
+                throw new UnsupportedOperationException("Only float dense vectors are supported for now");
```
We can work on this next, creating specific `BlockLoaders` for Byte and Bit field types.

carlosdelest commented on x-pack/plugin/esql/qa/testFixtures/src/main/java/org/elasticsearch/xpack/esql/CsvAssert.java:
```diff
@@ -145,6 +146,10 @@ private static void assertMetadata(
                     continue;
                 }
+                if (blockType == Type.DOUBLE && expectedType == DENSE_VECTOR) {
+                    // DENSE_VECTOR is internally represented as a double block
```
This could potentially change when we support byte and bit element types - we could create the appropriate blocks.

carlosdelest commented on x-pack/plugin/esql/qa/testFixtures/src/main/resources/mapping-default-incompatible.json:
```diff
@@ -63,18 +63,7 @@
             "type" : "keyword"
         },
         "salary_change": {
-            "type": "float",
```
Changing CSV loading made this a problem, as there were parsing exceptions when trying to index float numeric data into integers. I didn't see a convenient way out of this and decided to remove as this field is not being tested.

carlosdelest replied:
It is just changed for the `mapping-default-incompatible` mapping, which was created to test some incompatible field mappings that did not include subfields. I'll try to give this another shot but it will require changes to the CSV loader or the dataset 😢

fang-xing-esql replied:
> It is just changed for the `mapping-default-incompatible` mapping, which was created to test some incompatible field mappings that did not include subfields. I'll try to give this another shot but it will require changes to the CSV loader or the dataset 😢

Is it easier if we make another copy of employees's schema and data for `dense_vector` related tests?

Commit:
Fix test error when checking block loaders

Commit:
Fix value generation for dense vectors

Commit:
Merge remote-tracking branch 'carlosdelest/feature/esql_dense_vector_support' into feature/esql_dense_vector_support

Commit:
Fix testing for dense vectors and block loaders

Commit:
Remove dense_vector from CaseTests

Commit:
Provide support in tests for creating constant blocks from collection values

Commit:
Fix test error

Commit:
Fix unsupported types test

Commit:
Support synthetic source with non indexed vectors

Commit:
Remove unneeded code

Commit:
Add support for dense_vector in PositionToXContent and EsqlQueryResponseTests

Commit:
Fix unsupported types test

Commit:
Fix unsupported types test

Commit:
Fix block loader builder

elasticsearchmachine commented:
Pinging @elastic/es-analytical-engine (Team:Analytics)

Commit:
Ensure ordering when creating double blocks

ioanatia commented on x-pack/plugin/esql/qa/testFixtures/src/main/java/org/elasticsearch/xpack/esql/EsqlTestUtils.java:
```diff
@@ -827,6 +829,7 @@ public static Literal randomLiteral(DataType type) {
                     throw new UncheckedIOException(e);
                 }
             }
+            case DENSE_VECTOR -> Arrays.asList(randomArray(10, 10, i -> new Double[10], ESTestCase::randomDouble));
```
Should we actually `randomFloat()` and convert the value to double?
The rationale being that if we just use random double values, we might actually end up with something that can't be represented as a float and can't be used to actually index a `dense_vector` value.

carlosdelest replied:
Makes sense - Done in e8878a04db0264c7669fbd7daf12e0224cf4ed18

ioanatia commented on x-pack/plugin/esql/qa/testFixtures/src/main/resources/dense_vector.csv-spec:
```diff
@@ -0,0 +1,13 @@
+retrieveDenseVectorData
+required_capability: dense_vector_field_type
+
+FROM dense_vector
```
Can we add more csv tests here with commands we know should be supported 🙈 ?
We have `KEEP` already, but I'm thinking `DROP`, `RENAME` and simple `EVAL`s (`EVAL a = dense_vector_field`) might be supported already.
I know it should all just work - but for our own peace of mind it would be good to cover them.
Can be a single test that uses a combination of commands we know should be supported already.

carlosdelest replied:
I've added one in 93d45fc27731ff291b6a7feca0db24c7d04d6a0a. I'm not sure what this would catch, as the inner representation of dense_vector is a DoubleBlock and that is extensively tested for other fields.

I'm sure we will keep adding tests once we include arithmetic operators, conversion, etc to dense_vector.

carlosdelest replied:
> Do we expected that the items in a dense_vector have a fixed order?

Yes. It's crucial that the vector dimensions match between different vectors, so they can be compared.

>  a dense_vector looks very like a multi-valued double fields, it is hard to tell whether it is a dense_vector field or a double field with MV from its value looks

I've used MV as this seemed a supported way of internally representing a double array. From a user perspective, it's a different data type - it's a `dense_vector`, which will not necessarily be supported on the same functions, and it will always have the same number of dimensions / ordering for a specific mapping.

> I wonder what is the relationship between a double field with MV and a dense_vector.

We were thinking on adding a `TO_DENSE_VECTOR` cast function so users can specific a dense_vector using the MV double syntax:

`WHERE knn(field, TO_DENSE_VECTOR([0.1, 0.2, 0.3, ... , 1.0])`

Besides that, there should be no relation between the two. They are different data types that have the same representation (an array of elements).

> Do we expect the functions/commands that take multi-valued fields apply to dense_vector? Like those mv_xxx and to_xxx functions, mv_expand, stats by mv_fields etc.?

It would probably help to differentiate the two data types if `dense_vector` fields do not support MV functions.

We can provide support for multivalued functions, but most of them will not make sense in the context of a dense_vector (MV_APPEND, MV_CONCAT, MV_DEDUPE, MV_SORT, MV_SUM). Others can be useful even though they are not necessarily vector related (MV_COUNT, MV_FIRST,MV_MEDIAN, MV_MAX, MV_MIN), but supporting those could confuse ESQL users.

> dense_vector does not support sort or aggregation, 

Correct.

> does dense_vector support comparison, does it make sense to dense_vector fields?

We could support equality. Binary comparisons like greater / less than makes no sense for `dense_vector`.

ioanatia replied:
Thanks for adding the tests - this is addressed from my POV.
we can follow up on supporting more functions/operators - e.g. equality makes sense to me.

ioanatia commented on server/src/test/java/org/elasticsearch/index/mapper/vectors/DenseVectorFieldMapperTests.java:
```diff
@@ -2244,6 +2247,26 @@ protected boolean supportsEmptyInputArray() {
+    @Override
+    protected Function<Object, Object> loadBlockExpected(BlockReaderSupport blockReaderSupport, boolean columnReader) {
+        DenseVectorFieldType ft = (DenseVectorFieldType) blockReaderSupport.mapper().fieldType(blockReaderSupport.loaderFieldName());
+        if (ft.getElementType() != ElementType.FLOAT) {
+            return null;
+        }
+
+        return Function.identity();
```
nit (optional) `return ft.getElementType() == ElementType.FLOAT ? Function.identity() : null;`

carlosdelest replied:
I made the if stand out because we will get rid of that in the future 😉

Commit:
Changed randomDouble() for randomFloat() to generate test data

Commit:
Additional CSV tests

Commit:
Fix class casting

fang-xing-esql reviewed (commented):
Thank you @carlosdelest! At first, it looks to me that the value of a `dense_vector` field is very similar to a multi-valued `double` field, so I'm wondering if they work in similar ways, and when do we expect a `dense_vector` field works differently from the `double` field in ES|QL? I tried some queries to validate my thoughts, and a lot of times they work very similarly, and sometimes they don't. Here are my observations and you can have a look to see if they make sense. Perhaps I don't see an example of knn query yet, does knn query work on multi-valued `double` fields? I was looking for where a `dense_vector` field works differently from a multi-valued `double` field, and where it is expected to work the same as a multi-valued double `field`.

mapping
```
curl -u elastic:password -X PUT "localhost:9200/idx001?pretty" -H 'Content-Type: application/json' -d'
{
  "mappings": {
    "properties": {
      "numericfield": {"type": "double"},
      "mixedfield": {"type": "dense_vector","similarity": "l2_norm"}
    }
  }
}
'

curl -u elastic:password -X PUT "localhost:9200/idx002?pretty" -H 'Content-Type: application/json' -d'
{
  "mappings": {
    "properties": {
      "numericfield": {"type": "double"},
      "mixedfield": {"type" : "double"}
    }
  }
}
'

curl -X PUT "localhost:9200/idx001/_bulk?refresh&pretty" -H 'Content-Type: application/json' -d'
{"index": {}}
{"numericfield": 1,  "mixedfield" : [1.0, 2.0]}
'

curl -X PUT "localhost:9200/idx002/_bulk?refresh&pretty" -H 'Content-Type: application/json' -d'
{"index": {}}
{"numericfield": 2, "mixedfield" : [3.0, 4.0]}
'
```
queries 
```
+ curl -u elastic:password -v -X POST 'localhost:9200/_query?format=txt&pretty' -H 'Content-Type: application/json' '-d
{
  "query": "from idx00*"
}
'
  mixedfield   | numericfield  
---------------+---------------
null           |2.0            
null           |1.0            

+ curl -u elastic:password -v -X POST 'localhost:9200/_query?format=txt&pretty' -H 'Content-Type: application/json' '-d
{
  "query": "from idx00* | eval x = mixedfield::double"
}
'
{
  "error" : {
    "root_cause" : [
      {
        "type" : "verification_exception",
        "reason" : "Found 1 problem\nline 1:24: Cannot use field [mixedfield] due to ambiguities being mapped as [2] incompatible types: [dense_vector] in [idx001], [double] in [idx002]"
      }
    ],
    "type" : "verification_exception",
    "reason" : "Found 1 problem\nline 1:24: Cannot use field [mixedfield] due to ambiguities being mapped as [2] incompatible types: [dense_vector] in [idx001], [double] in [idx002]"
  },
  "status" : 400
}

+ curl -u elastic:password -v -X POST 'localhost:9200/_query?format=txt&pretty' -H 'Content-Type: application/json' '-d
{
  "query": "from idx001 | mv_expand mixedfield"
}
'
 numericfield  |  mixedfield   
---------------+---------------
1.0            |1.0            
1.0            |2.0            

+ curl -u elastic:password -v -X POST 'localhost:9200/_query?format=txt&pretty' -H 'Content-Type: application/json' '-d
{
  "query": "from idx001 | stats count(*) by mixedfield"
}
'
   count(*)    |  mixedfield   
---------------+---------------
1              |1.0            
1              |2.0            

+ curl -u elastic:password -v -X POST 'localhost:9200/_query?format=txt&pretty' -H 'Content-Type: application/json' '-d
{
  "query": "from idx001 | eval x = mv_sort(mixedfield)"
}
'
  mixedfield   | numericfield  |       x       
---------------+---------------+---------------
[1.0, 2.0]     |1.0            |[1.0, 2.0]     

+ curl -u elastic:password -v -X POST 'localhost:9200/_query?format=txt&pretty' -H 'Content-Type: application/json' '-d
{
  "query": "from idx001 | eval x = mixedfield::string"
}
'
{
  "error" : {
    "root_cause" : [
      {
        "type" : "verification_exception",
        "reason" : "Found 1 problem\nline 1:24: argument of [mixedfield::string] must be [aggregate_metric_double or boolean or cartesian_point or cartesian_shape or date_nanos or datetime or geo_point or geo_shape or ip or numeric or string or version], found value [mixedfield] type [dense_vector]"
      }
    ],
    "type" : "verification_exception",
    "reason" : "Found 1 problem\nline 1:24: argument of [mixedfield::string] must be [aggregate_metric_double or boolean or cartesian_point or cartesian_shape or date_nanos or datetime or geo_point or geo_shape or ip or numeric or string or version], found value [mixedfield] type [dense_vector]"
  },
  "status" : 400
}

+ curl -u elastic:password -v -X POST 'localhost:9200/_query?format=txt&pretty' -H 'Content-Type: application/json' '-d
{
  "query": "from idx002 | eval x= mixedfield::string"
}
'
  mixedfield   | numericfield  |       x       
---------------+---------------+---------------
[3.0, 4.0]     |2.0            |[3.0, 4.0]     
```

ioanatia commented on x-pack/plugin/esql/qa/testFixtures/src/main/java/org/elasticsearch/xpack/esql/EsqlTestUtils.java:
```diff
@@ -827,6 +829,7 @@ public static Literal randomLiteral(DataType type) {
                     throw new UncheckedIOException(e);
                 }
             }
+            case DENSE_VECTOR -> Arrays.asList(randomArray(10, 10, i -> new Double[10], () -> (double) randomFloat()));
```
this is on purpose, see https://github.com/elastic/elasticsearch/pull/126456#discussion_r2037397934

carlosdelest commented:
@fang-xing-esql both `dense_vector` and a `double` multivalued field share some common traits:
- They are composed as an array of `double` values
- Internal representation uses the multivalue mechanism in Blocks
- String representation should work the same way (unless we want it to be different)
- Users will use the multivalue format to specify them (potentially converting / casting to a `dense_vector` type)

They are different as in `dense_vector` fields:
- Values are always ordered
- They are not sortable
- Most MV_ functions make no sense to operate on dense_vector fields.
- dense_vector fields will have specific operations similar to script score existing [vector functions](https://www.elastic.co/guide/en/elasticsearch/reference/current/query-dsl-script-score-query.html#vector-functions).

I think what you observed makes sense for dense_vector fields. We may provide MV_ functions support, but in my mind that would make it confusing from a user perspective. `dense_vectors` should be treated as a single value, and not a collection of values.

Commit:
Use Float as block type for dense_vector

carlosdelest commented:
@ioanatia @fang-xing-esql I was able to use `FLOAT` blocks for `dense_vector` - changed that in cd462b843f82c05af6870407babbabb0af7b1955.

As Ioana noted, we're not publicly exposing `float` as a type here but `dense_vector`, so the change is doable as an implementation detail 👍 . Thanks!

Commit:
Fix tests

Commit:
[CI] Auto commit changes from spotless

Commit:
Fix test

Commit:
Merge remote-tracking branch 'carlosdelest/feature/esql_dense_vector_support' into feature/esql_dense_vector_support

ChrisHegarty commented on x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/lucene/ValuesSourceReaderOperator.java:
```diff
@@ -654,6 +654,16 @@ public BlockLoader.DoubleBuilder doubles(int expectedCount) {
 
+        @Override
+        public BlockLoader.FloatBuilder floatsFromDocValues(int expectedCount) {
+            return factory.newFloatBlockBuilder(expectedCount).mvOrdering(Block.MvOrdering.SORTED_ASCENDING);
```
I'm not quite sure, but the values are not SORTED for dense vectors, right?

ChrisHegarty commented on x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/lucene/ValuesSourceReaderOperator.java:
```diff
@@ -654,6 +654,16 @@ public BlockLoader.DoubleBuilder doubles(int expectedCount) {
+
+        @Override
+        public BlockLoader.FloatBuilder floats(int expectedCount) {
+            return factory.newFloatBlockBuilder(expectedCount);
```
I added support for float blocks/vectors to the compute engine quite a while ago, happy to see them being used now 👍

ChrisHegarty reviewed (commented):
I'm happy to see FloatBlocks being used, but they have too much flexibility - all different dims per position and also nulls. Maybe we just need an easy way/utility to assert their correct shape?

ioanatia reviewed (commented):
there's an outstanding [question](https://github.com/elastic/elasticsearch/pull/126456/files#r2046935314) from @ChrisHegarty that would be good to address  but otherwise LGTM

Commit:
Floats are not ordered for dense_vector values

Commit:
Use dense vector specific methods for creating builders, and check dimensions

Commit:
Avoid doing checks on dense vectors as we can't cast to BlockBuilder

Commit:
Merge remote-tracking branch 'origin/main' into feature/esql_dense_vector_support

# Conflicts:
#	x-pack/plugin/esql/qa/testFixtures/src/main/java/org/elasticsearch/xpack/esql/CsvTestsDataLoader.java
#	x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/action/EsqlCapabilities.java

carlosdelest commented:
> I'm happy to see FloatBlocks being used, but they have too much flexibility - all different dims per position and also nulls. Maybe we just need an easy way/utility to assert their correct shape?

@ChrisHegarty , I renamed some methods to make clear that we're creating dense vectors instead of floats, and added some checks for them, on c951ee770bfd6e5604d1ed4eae0121ff820f80bf.

I had to take some back on ba0a6b90247284b49180816f63e884fe6d6ed010, as I can't easily add a new float block builder given the sealed structure of builders. We may have to add a new Block type for `dense_vector`, maybe as a follow up?

LMKWYT

ioanatia commented on x-pack/plugin/esql/qa/testFixtures/src/main/resources/data/dense_vector.csv:
```diff
@@ -0,0 +1,3 @@
+id:l, vector:dense_vector
+0,    [1.0, 2.0, 3.0]
+1,    [4.0, 5.0, 6.0]
```
this might be very nitpicky - but can we add another `dense_vector` value that does not have ordered values?
this might be why we did not caught https://github.com/elastic/elasticsearch/pull/126456#discussion_r2046935314 during tests

carlosdelest replied:
mvOrdered does not impact retrieval as it's used as an optimization in some cases. But adding unordered data  uncovered a small fix that needed to be done to support multivalued style fields: 4b2126eeed19c6970ee60c4654eb92f75839f95f

ioanatia commented on x-pack/plugin/esql/src/internalClusterTest/java/org/elasticsearch/xpack/esql/DenseVectorFieldTypeIT.java:
```diff
@@ -0,0 +1,177 @@
+            .field("type", "dense_vector")
+            .field("index", index);
+        if (index) {
+            mapping.field("similarity", "l2_norm");
```
nit: is it possible to randomize the similarity option? otherwise we could leave this completely out, since it optional?

ioanatia commented on x-pack/plugin/esql/src/internalClusterTest/java/org/elasticsearch/xpack/esql/DenseVectorFieldTypeIT.java:
```diff
@@ -0,0 +1,177 @@
+
+        try (var resp = run(query)) {
+            assertColumnNames(resp.columns(), List.of("id", "vector"));
+            assertColumnTypes(resp.columns(), List.of("integer", "dense_vector"));
```
are we missing some assertions here for the values? to at least check that they are not nulls?

carlosdelest replied:
That's done in other tests - I started with this test to just check that the field types are retrieved correctly.

ioanatia commented on server/src/main/java/org/elasticsearch/index/mapper/BlockSourceReader.java:
```diff
@@ -303,6 +303,49 @@ public String toString() {
+
+        @Override
+        public String toString() {
+            return "BlockSourceReader.Floats";
```
Should this be `"BlockSourceReader.DenseVectors"`?

carlosdelest replied:
++, afe79f774a12cf0cd5c26fea871acb5736817f14

ioanatia commented on server/src/main/java/org/elasticsearch/index/mapper/BlockSourceReader.java:
```diff
@@ -303,6 +303,49 @@ public String toString() {
+
+        @Override
+        protected String name() {
+            return "Floats";
```
Again - just checking if this the value we want to return given that the class is called `DenseVectorBlockLoader`?

carlosdelest replied:
Good catch, I renamed this multiple times but didn't follow on this method. Thanks! afe79f774a12cf0cd5c26fea871acb5736817f14

Commit:
Add CSV tests with unordered data, and fix CSV data retrieval for that case

Commit:
Change source readers name() method

Commit:
Merge remote-tracking branch 'origin/main' into feature/esql_dense_vector_support

# Conflicts:
#	server/src/main/java/org/elasticsearch/index/mapper/vectors/DenseVectorFieldMapper.java
#	x-pack/plugin/esql/qa/testFixtures/src/main/java/org/elasticsearch/xpack/esql/CsvTestsDataLoader.java
#	x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/action/EsqlCapabilities.java

Commit:
[CI] Auto commit changes from spotless

carlosdelest commented:
Closing this PR, as the final approach will imply multiple field types for the different dense_vector element types.

Commit:
Merge remote-tracking branch 'origin/main' into feature/esql_dense_vector_support

# Conflicts:
#	x-pack/plugin/build.gradle
#	x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/data/BlockUtils.java
#	x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/action/EsqlCapabilities.java
#	x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/planner/PlannerUtils.java

Commit:
Spotless

Commit:
Merge remote-tracking branch 'carlosdelest/feature/esql_dense_vector_support' into feature/esql_dense_vector_support

carlosdelest commented:
@ioanatia @ChrisHegarty @fang-xing-esql I intend to merge this PR on Monday to provide initial support for dense_vector float element types. Please comment if you have any issue with that.

Next steps will be to work on KNN query and similarity functions, and eventually provide support for byte and bit element types.

Commit:
Fix tests

Commit:
Fix test suggested cast for joins

carlosdelest committed:
Merge branch 'main' into feature/esql_dense_vector_support

carlosdelest committed:
Merge branch 'main' into feature/esql_dense_vector_support

Commit:
[CI] Auto commit changes from spotless

Commit:
Fix LookupJoinTypesIT

Commit:
Merge remote-tracking branch 'carlosdelest/feature/esql_dense_vector_support' into feature/esql_dense_vector_support

elasticsearchmachine commented:
## 💔 Backport failed
| Status | Branch | Result |
|:------:|:------:|:------:|
| ❌ |  8.19  | Commit could not be cherrypicked due to conflicts |

You can use [sqren/backport](https://github.com/sqren/backport) to manually backport by running `backport --upstream elastic/elasticsearch --pr 126456`

carlosdelest commented:
## 💚 All backports created successfully

| Status | Branch | Result |
|:------:|:------:|:------|
|✅|8.19|[<img src="https://img.shields.io/github/pulls/detail/state/elastic/elasticsearch/128495">](https://github.com/elastic/elasticsearch/pull/128495)|

### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sorenlouv/backport)

<!--- Backport version: 9.5.1 -->
